contrib/buildinfo/src/BuildInfo.scala

```diff
@@ -15,33 +15,35 @@ trait BuildInfo extends ScalaModule {
     Map.empty[String, String]
   }
 
-  private def generateBuildInfo(members: Map[String, Any])(implicit dest: Ctx.Dest, log: Ctx.Log): Seq[PathRef] =
-    if (!members.isEmpty) {
-      val outputFile = dest.dest / "BuildInfo.scala"
+  def generatedBuildInfo: T[Seq[PathRef]] = T {
+    val logger: Logger = T.ctx.log
+    val members: Map[String, String] = buildInfoMembers()
+    if (members.nonEmpty) {
+      val outputFile = T.ctx.dest / "BuildInfo.scala"
       val internalMembers =
         members
           .map {
             case (name, value) => s"""  def ${name} = "${value}""""
           }
           .mkString("\n")
-      log.log.debug(s"Generating object [${buildInfoPackageName.map(_ + ".").getOrElse("")}${buildInfoObjectName}] with [${members.size}] members to [${outputFile}]")
+      logger.debug(s"Generating object [${buildInfoPackageName.map(_ + ".").getOrElse("")}${buildInfoObjectName}] with [${members.size}] members to [${outputFile}]")
       os.write(
         outputFile,
-        s"""|${buildInfoPackageName.map(p => s"package ${p}").getOrElse("")}
-          |object ${buildInfoObjectName} {
-          |$internalMembers
-          |}""".stripMargin
+        s"""|${buildInfoPackageName.map(packageName => s"package ${packageName}\n").getOrElse("")}
+            |object ${buildInfoObjectName} {
+            |$internalMembers
+            |}""".stripMargin
       )
       Seq(PathRef(outputFile))
     } else {
-      log.log.debug("No build info member defined, skipping code generation")
+      logger.debug("No build info member defined, skipping code generation")
       Seq.empty[PathRef]
     }
-
-  def buildInfo = T {
-    generateBuildInfo(buildInfoMembers())
   }
 
-  override def generatedSources: Target[Seq[PathRef]] = T { super.generatedSources() ++ buildInfo() }
+  override def generatedSources = T {
+    super.generatedSources() ++
+      generatedBuildInfo().map(pathRef => PathRef(pathRef.path / os.up))
```

Comment:
Looks like "guessing" the best directory here. What if `generateBuildInfo` also generates sub-directories for each package?

Reply:
Right, but I haven't found a better way to specify a folder. Do you have any ideas how to find and  "specify" the 'dest/ folder here?

Reply:
What about returning a `(PathRef, Seq[PathRef])` instead, the first is the directory, the second the generated resources. That way, when generating nothing, the dir is still present, but the seq might be empty.

Reply:
By "returning" I refer to the `generateBuildInfo` target, of course.

Reply:
Will change it when I'm back from work

Reply:
I was thinking, maybe one can use `myTask.ctx().dest` and indeed there is such a thing as `generatedBuildInfo.ctx` (notice the absense of brackets) of type `mill.define.NamedTask#ctx`. This property in turn is of type `mill.define.Ctx`, which contains a lot of different properties, but not `dest`. What do you think, does it make sense to add the dest property there? So that it can be used and referenced by others. Or is there a better place for this?

I will implement it for now as you suggested, by returning a tuple. But for the future this could be a solution as well. Because returning and destructing tuple is not the nicest thing, considering that the task should return only one value.

Reply:
Imagine, someone is overriding `generateBuildInfo`, then of course, "dest" might be located somewhere else, so you must dynamically bind to the actual target or better, use the return value.

In fact, a tuple IS one return value. But if you prefer it you can use a case class instead (`case class Generated(dir: Path, files: Seq[Path])`).

Reply:
I know, I tried it with case classes as well.

The only thing is that you have to wrap Path into a PathRef to have it serializable.

Latest changes already pushed.